documentation/specs/src/interoperability/ethereum-bridge/proofs.md

```diff
@@ -0,0 +1,96 @@
+Since the Ethereum smart contract should accept any bridge
+header signed by 2 / 3 of the staking validators, it needs up-to-date
+knowledge of:
+- The current validators' public keys
+- The current stake of each validator
+
+This means that by the end of every Namada epoch, a special transaction must be
+sent to the Ethereum smart contracts detailing the new public keys and stake
+of the new validator set. This message must also be signed by at least 2 / 3
+of the current validators as a "transfer of power".
+
+If vote extensions are available, this signed data can be constructed
+using them. Otherwise, validators must send protocol txs to be included on
+the ledger. Once a quorum exist on chain, they can be aggregated into a
+single message that can be relayed to Ethereum. Signing an
+invalid  validator transition set will be considered a slashable offense.
+
+Due to asynchronicity concerns, this message should be submitted well in
+advance of the actual epoch change. It should happen at the beginning of each
+new epoch. Bridge headers to ethereum should include the current Namada epoch
+so that the smart contract knows how to verify the headers. In short, there
+is a pipelining mechanism in the smart contract - the active validators for epoch `n` submit details of the active validator set for epoch `n+1`.
+
+Such a message is not prompted by any user transaction and thus will have
+to be carried out by a _bridge relayer_. Once the necessary data to 
+construct the transfer of power  message is on chain, any time afterwards a 
+Namada bridge process may take it to craft the appropriate header to the 
+Ethereum smart contracts.
+
+The details on bridge relayers are below in the corresponding section.
+
+Signing incorrect headers is considered a slashable offense. Anyone witnessing
+an incorrect header that is signed may submit a complaint (a type of transaction)
+to initiate slashing of the validator who made the signature.
+
+## Namada Bridge Relayers
+
+Validator changes must be turned into a message that can be communicated to
+smart contracts on Ethereum. These smart contracts need this information
+to verify proofs of actions taken on Namada.
+
+Since this is protocol level information, it is not user prompted and thus
+should not be the responsibility of any user to submit such a transaction.
+However, any user may choose to submit this transaction anyway.
+
+This necessitates a Namada node whose job it is to submit these transactions on
+Ethereum by the conclusion of each Namada epoch. This node is called the
+__bridge relayer__. In theory, since this message is publicly available
+on the blockchain, anyone can submit this transaction, but only the
+bridge relayer will be directly compensated by Namada.
+
+The bridge relayer will be chosen to be the proposer of the first block of the 
+new epoch. Anyone else may relay this message, but must pay for the fees out of
+their own pocket.
```

Comment:
How about letting a block proposer choose whether they want to be the bridge relayer by injecting a transfer of power message into their block? (https://github.com/anoma/namada/pull/393#issuecomment-1232809428 - "Submission of the transfer of power message onto Namada chain").

We should add more details of how/when the transfer of power message is submitted on chain to Namada and who actually submits it. e.g. maybe anyone can craft this transfer of power message and submit it on chain as a normal transaction, or maybe it needs to be a validator or the block proposer. It should happen asap after there are enough signatures on chain of the next epoch's validator set.

Reply:
The problem is that whoever does it gets paid. So the protocol should specify who gets paid, otherwise we incentivize other validators to reject valid proposals to get funds for themselves.

Reply:
The latter part is easy, there is just a scattered bunch of protocol txs on chain with a `Signed<ValidatSet>` struct inside. Someone just collects all those together and submits to ethereum. I can add some struct defs. The when is as soon as possible, but it is impossible to enforce a timeline.

Reply:
Validators shouldn't reject valid blocks (including ones that have a valid transfer of power message injected by the block proposer), if they do so they should be slashed. As a more general issue, I think validators should have to provide a signed rejection of a block proposal that could later be submitted as evidence, but I'm guessing we don't have anything like this implemented yet. 

If we go with the bridge relayer being designated as the first block proposer of an epoch, I guess they should be naturally incentivized to submit the transfer of power message on chain as soon as it is possible. We should add details to the spec of what this transfer of power will look like

Reply:
Adding slashing would fix the problem, it's just a lot more engineering effort. So I'm more inclined to have the protocol specify who gets paid.

Reply:
> The latter part is easy, there is just a scattered bunch of protocol txs on chain with a `Signed<ValidatSet>` struct inside. Someone just collects all those together and submits to ethereum. I can add some struct defs. The when is as soon as possible, but it is impossible to enforce a timeline.

(submitted my last comment before seeing this)

I might be misunderstanding the spec, can the transfer of power be crafted offchain arbitrarily and submitted, without first being submitted onto Namada?

Reply:
Yes, I will clarify this.

Reply:
ok, that was a big misunderstanding on my part, I've been reading this part of the spec as if there was this step where we submit a transfer of power message onto Namada first